lib/graphs/digraph.nit

```diff
@@ -0,0 +1,606 @@
+						queue.add(wp)
+						dist[wp] = dist[w] + 1
+					end
+				end
+			end
+			if queue.is_empty then break
+		end
+		return dist.get_or_null(v)
+	end
+
+	# -------------------- #
+	# Connected components #
+	# -------------------- #
+
+	# Returns the weak connected components of this digraph.
+	#
+	# The weak connected components of a digraph are the usual
+	# connected components of its associated undirected graph,
+	# i.e. the graph obtained by replacing each arc by an edge.
+	fun weak_connected_components: DisjointSet[V]
+	do
+		var components = new DisjointSet[V]
+		components.add_all(vertices)
+		for arc in arcs do
+			components.union(arc.source, arc.target)
+		end
+		return components
+	end
+
+	# Returns the strongly connected components of this digraph.
+	#
+	# Two vertices `u` and `v` belong to the same strongly connected
+	# component if and only if there exists a path from `u` to `v`
+	# and there exists a path from `v` to `u`.
+	#
+	# This is computed in linear time (Tarjan's algorithm).
+	fun strongly_connected_components: DisjointSet[V]
+	do
+		sccs = new DisjointSet[V]
+		sccs.add_all(vertices)
+		tarjan_index = 0
+		tarjan_stack = (new Array[V]).as_lifo
+		tarjan_vertex_to_index = new HashMap[V, Int]
+		tarjan_ancestor = new HashMap[V, Int]
+		tarjan_in_stack = new HashMap[V, Bool]
+		for v in vertices do
+			tarjan(v)
+		end
+		return sccs
+	end
+
+	# The recursive part of Tarjan's algorithm
+	private fun tarjan(u: V)
+	do
+		tarjan_vertex_to_index[u] = tarjan_index
+		tarjan_ancestor[u] = tarjan_index
+		tarjan_index += 1
+		tarjan_stack.add(u)
+		tarjan_in_stack[u] = true
+		for v in successors(u) do
+			if not tarjan_vertex_to_index.keys.has(v) then
+				tarjan(v)
+				tarjan_ancestor[u] = tarjan_ancestor[u].min(tarjan_ancestor[v])
+			else if tarjan_in_stack[v] then
+				tarjan_ancestor[u] = tarjan_ancestor[u].min(tarjan_vertex_to_index[v])
+			end
+		end
+		if tarjan_vertex_to_index[u] == tarjan_ancestor[u] then
+			var v
+			loop
+				v = tarjan_stack.take
+				tarjan_in_stack[v] = false
+				sccs.union(u, v)
+				if u == v then break
+			end
+		end
+	end
+
+	# The strongly connected components computed in Tarjan's algorithm
+	private var sccs: DisjointSet[V] is noinit
+	# An index used for Tarjan's algorithm
+	private var tarjan_index: Int is noinit
+	# A stack used for Tarjan's algorithm
+	private var tarjan_stack: Queue[V] is noinit
+	# A map associating with each vertex its index
+	private var tarjan_vertex_to_index: HashMap[V, Int] is noinit
+	# A map associating with each vertex its ancestor in Tarjan's algorithm
+	private var tarjan_ancestor: HashMap[V, Int] is noinit
+	# True if and only if the vertex is in the stack
+	private var tarjan_in_stack: HashMap[V, Bool] is noinit
+end
+
+# A directed graph represented by hash maps
+class HashMapDigraph[V: Object]
+	super AbstractDigraph[V]
+
+	# Attributes
+	private var pred_map = new HashMap[V, Array[V]]
+	private var succ_map = new HashMap[V, Array[V]]
```

Comment:
Personally, I would have reified the node with a private structure to factorize things.

~~~nit
private var nodes = new HashMap[V, HashMapNode[[V]]
#...
private class HashMapNode [V]
   var preds = new Array[V]
   var succs = new Array[V]
end
~~~

This add an indirection level but has the advantage to merge the two hashmaps.
Another benefit is that you have a private class for nodes that you can use to store other data if you want.

So there is pros and cons. I do not know it this worth the change, or even some objective measure of the cost of each one.